### PR TITLE
CSCQVAIN-288, 308, 297: rewrite of date, time and date time pickers

### PR DESCRIPTION
**Pull request by jppiiroinen**
This PR adds support for free-form date, time and timezone/offset entries.

What is missing still is the helpers, such as visual date picker and timezone/offset picker. But I will do those in separate PR for this same issue.

### CODE REVIEW

File: src/components/Date.vue [hunk 1]
```diff
@@ -9,13 +9,7 @@
...
 			<div class="row">
 				<datepicker
-					:inline="false"
-					placeholder="Select Date"
-					monday-first
-					bootstrapStyling
-					format="dd.MM.yyyy"
-					:id="property + '_date'"
-					class="col"
+					v-bind:format="schema.format"
```
Comment (Kauhia) [inline src/components/Date.vue]: Why v-bind:

File: src/components/Date.vue [hunk 2]
```diff
@@ -55,40 +49,16 @@ export default {
...
-		},
 	},
 	created() {
-		this.date = this.value ? new Date(this.value) : null
+		this.date = this.value ? this.value : null
 	},
 	watch: {
 		date() {
-			this.$store.commit('updateValue', { p: this.parent, prop: this.property, val: this.dateString })
+			console.log("date changed")
```
Comment (Kauhia) [inline src/components/Date.vue]: console.log

File: src/components/DatePicker.vue [hunk 3]
```diff
@@ -0,0 +1,113 @@
+<!-- ADD_LICENSE_HEADER -->
+<template>
+    <div class="input-group">
+        <div class="input-group-prepend">
+            <span class="input-group-text">Date</span>
+        </div>
+        <input
+            :key="componentKey"
+            v-bind:value="dateString"
+            v-on:input="validateDate"
```
Comment (Kauhia) [inline src/components/DatePicker.vue]: v-on:input or @input, I prefer @input just because we used it before.

File: src/components/DatePicker.vue [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+            placeholder="dd.MM.yyyy"
+            class="form-control" />
+    </div>
+</template>
+
+<script>
+
+export default {
+	name: 'datepicker',
+	components: {
```
Comment (Kauhia) [inline src/components/DatePicker.vue]: Not needed

File: src/components/DatePicker.vue [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+    },
+    model: {
+        prop: 'value',
+        event: 'input'
+    },
+    props: {
+        format: String,
+        value: String,
+    },
+	data: function () {
```
Comment (Kauhia) [inline src/components/DatePicker.vue]: data() {} should be enough

File: src/components/DatePicker.vue [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+            internal_value: null,
+            initial_value: null,
+            componentKey: 0,
+            isInitializing: true,
+            dateRegexp: /^((([1-9]|0[1-9]|[12]\d|3[01])\.([1-9]|0[1-9]|1[0-2])(\.[12]\d{3})?)|([1-9]|0[1-9]|[12]\d|3[01]))$/
+        }
+	},
+	computed: {
+        dateString: {
+			get: function() {
```
Comment (Kauhia) [inline src/components/DatePicker.vue]: Same for all functions, no function keyword needed

File: src/components/DatePicker.vue [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+            var new_value = event.target.value
+            if (!new_value || new_value === '') {
+                this.internal_value = new_value
+                this.initial_value = new_value
+                this.$emit('input', new_value)
+            } else if (this.dateRegexp.test(new_value)) {
+                var hours = new_value.split(".")
+                var parts = hours.length
+
+                for(var i=0; i<hours.length; i++) {
```
Comment (Kauhia) [inline src/components/DatePicker.vue]: Just coding style things: would prefer functional style here

File: src/components/DatePicker.vue [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+                    hours[i] = hours[i].padStart(2, '0')
+                }
+                for (var i=0; i<3-parts; i++) {
+                    hours.push("00")
+                }
+
+                // ensure that the month is not empty
+                if (hours[1] === "00") {
```
Comment (Kauhia) [inline src/components/DatePicker.vue]: Would these checks be better if made with date-fns?

File: src/components/DatePicker.vue [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+
+                new_value = hours.join(".")
+
+                this.$emit('input', new_value)
+                this.internal_value = new_value
+                this.initial_value = new_value
+            } else {
+                this.internal_value = this.initial_value
+            }
+            this.forceRerender()
```
Comment (Kauhia) [inline src/components/DatePicker.vue]: Why is this needed?

File: src/components/Date.vue (hunk 2 above)
Comment (tahme) [inline src/components/Date.vue]: remove console.log

File: src/components/DatePicker.vue [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+        }
+	},
+	created() {
+        this.isInitializing = true
+        this.internal_value = this.value
+        this.initial_value = this.value
+        this.isInitializing = false
+	}
+}
+</script>
```
Comment (tahme) [inline src/components/DatePicker.vue]: add newline

File: src/components/DateTimePicker.vue [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+            if (this.isInitializing) { return }
+            this.updateValue()
+		},
+		date() {
+            if (this.isInitializing) { return }
+            this.updateValue()
+		},
+	},
+}
+</script>
```
Comment (tahme) [inline src/components/DateTimePicker.vue]: missing newline at eof

File: src/components/TimePicker.vue [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+        }
+	},
+	created() {
+        this.isInitializing = true
+        this.internal_value = this.value
+        this.initial_value = this.value
+        this.isInitializing = false
+	}
+}
+</script>
```
Comment (tahme) [inline src/components/TimePicker.vue]: missing newline at eof

File: src/components/TimeZonePicker.vue [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+        }
+	},
+	created() {
+        this.isInitializing = true
+        this.internal_value = this.value
+        this.initial_value = this.value
+        this.isInitializing = false
+	}
+}
+</script>
```
Comment (tahme) [inline src/components/TimeZonePicker.vue]: missing newline at eof

File: src/components/TimePicker.vue [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+	},
+	computed: {
+        timeString: {
+			get: function() {
+                return this.internal_value
+			},
+			set: function(new_value) {
+                if (this.isInitializing) { return }
+				if (this.timeRegexp.test(new_value)) {
+                    this.internal_value = new_value
```
Comment (tahme) [inline src/components/TimePicker.vue]: weird indentation here

File: src/components/TimeZonePicker.vue [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+            timezoneRegexp: /^[+-]?((0[0-9]|1\d|2[0-3]|[0-9])|((0[0-9]|1\d|2[0-3]|[0-9]):){1})$/
+        }
+	},
+	computed: {
+        timezoneString: {
+			get: function() {
+                return this.internal_value
+			},
+			set: function(new_value) {
+                if (this.isInitializing) { return }
```
Comment (tahme) [inline src/components/TimeZonePicker.vue]: weird indentation here

File: src/components/DateTimePicker.vue [hunk 16]
```diff
@@ -0,0 +1,121 @@
+<!-- ADD_LICENSE_HEADER -->
+<template>
+    <b-container class="datetimepicker" :key="componentKey">
```
Comment (Kauhia) [inline src/components/DateTimePicker.vue]: why :key

File: src/components/DateTimePicker.vue [hunk 17]
```diff
@@ -0,0 +1,121 @@
...
+        </b-row>
+    </b-container>
+</template>
+
+<style>
+
+</style>
+
+<script>
+//import Datepicker from 'vuejs-datepicker'
```
Comment (Kauhia) [inline src/components/DateTimePicker.vue]: remove

File: src/components/DateTimePicker.vue [hunk 18]
```diff
@@ -0,0 +1,121 @@
...
+        "timepicker": TimePicker,
+        "timezonepicker": TimeZonePicker,
+    },
+    props: {
+        format: String,
+        value: String,
+        title: String,
+        description: String,
+    },
+	data: function () {
```
Comment (Kauhia) [inline src/components/DateTimePicker.vue]: function syntax

File: src/components/DatePicker.vue [hunk 19]
```diff
@@ -0,0 +1,113 @@
...
+        format: String,
+        value: String,
+    },
+	data: function () {
+        return {
+            internal_value: null,
+            initial_value: null,
+            componentKey: 0,
+            isInitializing: true,
+            dateRegexp: /^((([1-9]|0[1-9]|[12]\d|3[01])\.([1-9]|0[1-9]|1[0-2])(\.[12]\d{3})?)|([1-9]|0[1-9]|[12]\d|3[01]))$/
```
Comment (Kauhia) [inline src/components/DatePicker.vue]: Please comment this so that people who are bad with regex (like me..) can get the logic

File: src/components/TimeZonePicker.vue [hunk 20]
```diff
@@ -0,0 +1,111 @@
...
+                this.$emit('input', new_value)
+            } else if (this.timezoneRegexp.test(new_value)) {
+                var symbol = "-"
+                if (new_value[0]==="+" || new_value[0]==="-") {
+                    symbol = new_value[0]
+                    new_value = new_value.substring(1)
+                }
+                var hours = new_value.split(":")
+                var parts = hours.length
+                console.log(hours)
```
Comment (tahme) [inline src/components/TimeZonePicker.vue]: remove console.log

File: src/components/TimeZonePicker.vue [hunk 21]
```diff
@@ -0,0 +1,111 @@
...
+			set: function(new_value) {
+                if (this.isInitializing) { return }
+				if (this.timezoneRegexp.test(new_value)) {
+                    this.internal_value = new_value
+				}
+			}
+        },
+	},
+	methods: {
+        submitChange: function(event) {
```
Comment (Kauhia) [inline src/components/TimeZonePicker.vue]: Style thing: In other components changes are emited as internal value changes (watch), in this it is on change. I (currently, my style is in a bit flux) like the idea that watch is used to update parent. It moves the update logic dependency from template to the js part that in my opinion makes it easier to follow the data flow. In theory, props --> template, template event --> internal state, internal state --> update parent (or store). Hopefully this doesn't cause confusion, I'm not rely sure I delivered the reasoning well. And in any case this is not a change request but rather a note on style.

**Review by Kauhia — CHANGES_REQUESTED** [review]

File: src/components/DatePicker.vue [hunk 22]
```diff
@@ -0,0 +1,113 @@
...
+					var date_array = new_value.split(".")
+                    this.internal_value = date_array[2] + "-" + date_array[1] + "-" + date_array[0]
+				}
+			}
+        },
+	},
+	methods: {
+        submitChange: function(event) {
+            if (this.isInitializing) { return }
+            var new_value = event.target.value
```
Comment (tahme) [inline src/components/DatePicker.vue]: prefer `let` instead of `var`

File: src/components/DatePicker.vue [hunk 23]
```diff
@@ -0,0 +1,113 @@
...
+                this.internal_value = new_value
+                this.initial_value = new_value
+            } else {
+                this.internal_value = this.initial_value
+            }
+            this.forceRerender()
+        },
+        validateDate: function(event) {
+            if (this.isInitializing) { return }
+            var new_value = event.target.value
```
Comment (tahme) [inline src/components/DatePicker.vue]:
```suggestion
            const new_value = event.target.value
```

File: src/components/DatePicker.vue (hunk 19 above)
Comment (jppiiroinen) [reply src/components/DatePicker.vue]: I will write a unit tests for all these in the next PR.

File: src/components/TimeZonePicker.vue (hunk 21 above)
Comment (jppiiroinen) [reply src/components/TimeZonePicker.vue]: I can change this in the next PR when I add the helpers.

File: src/components/DatePicker.vue (hunk 23 above)
Comment (jppiiroinen) [reply src/components/DatePicker.vue]: I did not use those as I was thinking that not all browsers supports them. but I will take them into use

**Comment by jppiiroinen** [comment]
The code is now broken. needs to be fixed.

File: src/components/DatePicker.vue (hunk 23 above)
Comment (tahme) [reply src/components/DatePicker.vue]: babel can handle those, so they should be safe to use.

File: src/components/TimeZonePicker.vue (hunk 21 above)
Comment (jppiiroinen) [reply src/components/TimeZonePicker.vue]: might be improved already

File: src/components/DatePicker.vue [hunk 24]
```diff
@@ -0,0 +1,125 @@
...
+					var date_array = new_value.split(".")
+					this.internal_value = date_array[2] + "-" + date_array[1] + "-" + date_array[0]
+				}
+			}
+		},
+	},
+	methods: {
+		submitChange(event) {
+			if (this.isInitializing) { return }
+			var new_value = event.target.value
```
Comment (Kauhia) [inline src/components/DatePicker.vue]: const

File: src/components/Date.vue [hunk 25]
```diff
@@ -9,13 +9,7 @@
 		<div slot="input" class="container">
```
Comment (tahme) [inline src/components/Date.vue]: I think this
```
		<div slot="input" class="container">
				<div slot="input" class="row">
```
should be
```
		<div slot="input" class="row">
			<div class="col">
```

because currently the classes go:
```
div.container (from RecordField)
  div.container (from Date)
    div.row
      div.datetimepicker.container
         ...
```

File: src/components/DatePicker.vue [hunk 26]
```diff
@@ -0,0 +1,125 @@
...
+
+				// ensure that the year is not empty
+				hours[2] = hours[2].padStart(4, '0')
+				if (hours[2] === "0000") {
+					hours[2] = new Date().getFullYear()
+				}
+
+				new_value = hours.join(".")
+
+				this.$emit('input', new_value)
```
Comment (Kauhia) [inline src/components/DatePicker.vue]: emit before updating internal state might cause problems if something in parent is changed and that change is propagated to child component that then is rerendered

File: src/components/TimePicker.vue [hunk 27]
```diff
@@ -0,0 +1,99 @@
...
+		return {
+			internal_value: null,
+			initial_value: null,
+			isInitializing: true,
+			inputRegexp: /^[\d:]+$/,
+			timeRegexp: /^((0[0-9]|1\d|2[0-3]|[0-9])|((0[0-9]|1\d|2[0-3]|[0-9]):){1}(([0-5][0-9]|[0-9]):?){1,2})$/
+		}
+	},
+	computed: {
+		timeString: {
```
Comment (Kauhia) [inline src/components/TimePicker.vue:39]: function syntax

File: src/components/TimePicker.vue [hunk 28]
```diff
@@ -0,0 +1,99 @@
...
+		},
+		validate(event) {
+			if (this.isInitializing) { return }
+			var new_value = event.target.value
+			if (this.timeRegexp.test(new_value)) {
+				this.internal_value = new_value
+				this.initial_value = this.internal_value
+			} else if (!this.inputRegexp.test(new_value) && new_value !== '') {
+				this.internal_value = this.initial_value
+				this.$forceUpdate();
```
Comment (Kauhia) [inline src/components/TimePicker.vue:85]: Why is forceUpdate required?

File: src/components/TimeZonePicker.vue [hunk 29]
```diff
@@ -0,0 +1,148 @@
...
+				if (this.isInitializing) { return }
+				if (this.timezoneRegexp.test(new_value)) {
+					this.internal_value = new_value
+				}
+			}
+		},
+	},
+	methods: {
+		updateOffset() {
+			// this.date is in day.month.year format
```
Comment (Kauhia) [inline src/components/TimeZonePicker.vue]: I have a feeling that there is something dangerous in here since date things are calculated without a library and that always goes wrong eventually..

File: src/components/TimeZonePicker.vue [hunk 30]
```diff
@@ -0,0 +1,148 @@
...
+			} else if (is_utc) {
+				symbol = "+"
+			} else {
+				symbol = "-"
+			}
+			this.initial_value = symbol + hours + ":" + minutes
+			this.internal_value = this.initial_value
+			this.$emit('input', this.initial_value)
+		},
+		submitChange(event) {
```
Comment (Kauhia) [inline src/components/TimeZonePicker.vue:82]: Is this the same logic as in the another component? Maybe share it with mixin?

**Review by Kauhia — DISMISSED** [review]

File: src/components/Date.vue (hunk 25 above)
Comment (jppiiroinen) [reply src/components/Date.vue]: fixed

File: src/components/TimePicker.vue (hunk 27 above)
Comment (jppiiroinen) [reply src/components/TimePicker.vue:39]: that is computed setter/getter

File: src/components/TimeZonePicker.vue (hunk 29 above)
Comment (jppiiroinen) [reply src/components/TimeZonePicker.vue]: noted and ignored

File: src/components/TimeZonePicker.vue (hunk 30 above)
Comment (jppiiroinen) [reply src/components/TimeZonePicker.vue:82]: yes we should do that, but there is no time for a proper refactoring tour for the release

File: src/components/TimePicker.vue (hunk 28 above)
Comment (jppiiroinen) [reply src/components/TimePicker.vue:85]: so that the computed getter gets called

File: src/components/DatePicker.vue (hunk 26 above)
Comment (jppiiroinen) [reply src/components/DatePicker.vue]: noted.. I will fix this

**Review by tahme — APPROVED** [review]

File: src/components/TimeZonePicker.vue [hunk 31]
```diff
@@ -0,0 +1,154 @@
...
+<template>
+	<div class="input-group">
+		<div class="input-group-prepend">
+			<b-button
+				@click="updateOffset"> 
+				UTC ±
+			</b-button>
+		</div>
+		<input
+			v-bind:value="timezoneString"
```
Comment (jppiiroinen) [inline src/components/TimeZonePicker.vue]: here is one v-bind left
